### PR TITLE
Changes to support Query insights for MYSQL

### PR DESCRIPTION
**Pull request by cedric66**
Upgraded "GoogleCloudPlatform/sql-db/google//modules/mysql" module version to from 9.0.0 to 13.0.0 to support Query insights for MYSQL
After version 10.0.0 allocated_ip_range must now be specified for instances. Added allocated_ip_range = null variable

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -194,6 +194,11 @@ variable "public_access_master_instance" {
   default     = false
 }
 
+variable "allocated_ip_range" {
+  description = "For MYSQL db, adding this property to Cloud SQL modules will allow users to select a specific allocated range for their private instances."
+  type        = string
+  default     = null
+}
```
Comment (syedrakib) [inline variables.tf:229]: Keep an empty line between 2 varialbes

File: variables.tf [hunk 2]
```diff
@@ -194,6 +194,11 @@ variable "public_access_master_instance" {
   default     = false
 }
 
+variable "allocated_ip_range" {
+  description = "For MYSQL db, adding this property to Cloud SQL modules will allow users to select a specific allocated range for their private instances."
```
Comment (syedrakib) [inline variables.tf]: This is mysql_db module. So we know it's for MySQL.

```suggestion
  description = "Allow users to select a specific IP range for their private instances."
```

Also, what would be the recommended IP width for this? `/16`? `/20`? `/24`? `/32`.

File: variables.tf [hunk 3]
```diff
@@ -286,3 +291,21 @@ variable "maintenance_window" {
     update_track = "stable"
   }
 }
+
+# ----------------------------------------------------------------------------------------------------------------------
+# To enable Query Insights
+# ----------------------------------------------------------------------------------------------------------------------
```
Comment (syedrakib) [inline variables.tf]: The variable name is pretty clear here. No need to populate more comments for it.

In fact, if we want to clarify the purpose of the variable, we can do so in the description of the variable itself - which will also get included into the documentation pages that are generated by terraform. So that's cleaner than adding comments.

File: variables.tf [hunk 4]
```diff
@@ -286,3 +291,21 @@ variable "maintenance_window" {
...
+    query_string_length     = number
+    record_application_tags = bool
+    record_client_address   = bool
+  })
+  default = {
+    query_string_length     = 1024
+    record_application_tags = false
+    record_client_address   = false
+  }
+}
```
Comment (syedrakib) [inline variables.tf]: Kindly add a newline at the end of the file.

VSCode actually makes it easy by enabling an option to auto-add a newline at the end of every file.

File: main.tf [hunk 5]
```diff
@@ -43,7 +43,7 @@ resource "google_project_service" "cloudsql_api" {
 
 module "google_mysql_db" {
   source                          = "GoogleCloudPlatform/sql-db/google//modules/mysql"
-  version                         = "9.0.0"
+  version                         = "13.0.0"
```
Comment (syedrakib) [inline main.tf]: Has this been tested in dev/stg if it will end up breaking anything else? Need to test this out before we can merge this.

You can follow this https://www.dailytask.co/task/test-your-terraform-module-before-publishing-it-ahmed-zidan to learn how to test an unpublished terraform module using its branch name. It's written by @ahmedzidan

File: main.tf [hunk 6]
```diff
@@ -89,7 +90,11 @@ module "google_mysql_db" {
     retained_backups               = null
     retention_unit                 = null
   }
-
+  insights_config = {
+    query_string_length     = var.insights_config.query_string_length
+    record_application_tags = var.insights_config.record_application_tags
+    record_client_address   = var.insights_config.record_client_address
+  }
```
Comment (syedrakib) [inline main.tf]: Would suggest to just use `insights_config = var.insights_config`

Whatever keys need to be passed here will be flexible from the variable itself. Much like what was done here https://github.com/airasia/terraform-google-postgres_db/pull/6 for our postgres_db module

**Review by syedrakib — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 1 above)
Comment (cedric66) [reply variables.tf:229]: done

File: variables.tf (hunk 3 above)
Comment (cedric66) [reply variables.tf]: done

File: variables.tf (hunk 4 above)
Comment (cedric66) [reply variables.tf]: done

File: main.tf (hunk 6 above)
Comment (cedric66) [reply main.tf]: done

File: variables.tf [hunk 7]
```diff
@@ -194,6 +194,12 @@ variable "public_access_master_instance" {
   default     = false
 }
 
+variable "allocated_ip_range" {
+  description = "Allow users to select a specific IP range for their private instances"
```
Comment (syedrakib) [inline variables.tf]: What would be the recommended IP width for this? /16? /20? /24? /32. We should recommend something here like we have recommended for `ip_ranges` variable in the [vpc_network](https://registry.terraform.io/modules/airasia/vpc_network/google/latest?tab=inputs) module.

File: main.tf (hunk 5 above)
Comment (cedric66) [reply main.tf]: Hi Rakib, this has been tested in development and it works as expected

File: variables.tf (hunk 7 above)
Comment (cedric66) [reply variables.tf]: Added the cidr range information

File: main.tf [hunk 8]
```diff
@@ -103,13 +104,16 @@ module "google_mysql_db" {
...
-      disk_size           = var.disk_size_gb_read_replica
-      disk_type           = "PD_SSD"
-      user_labels         = var.labels_read_replica
-      encryption_key_name = null
+      database_flags        = local.db_flags_read_replica
+      disk_autoresize       = var.disk_auto_resize_read_replica
+      disk_autoresize_limit = var.disk_autoresize_limit
+      disk_size             = var.disk_size_gb_read_replica
+      disk_type             = "PD_SSD"
+      availability_type     = var.read_replica_availability_type
```
Comment (mfarix) [inline main.tf]: Make the variable name as `availability_type_read_replica` to make it consistent with variable naming convention.

Comment (mfarix) [reply main.tf]: Or better yet, follow the way it is done on master node. Something like:
`var.highly_available_read_replica ? "REGIONAL" : "ZONAL"`

File: variables.tf [hunk 9]
```diff
@@ -194,6 +210,12 @@ variable "public_access_master_instance" {
   default     = false
 }
 
+variable "allocated_ip_range" {
+  description = "Allow users to select a specific IP range for their private instances. private_g_services: A CIDR range (/20 advised) for Google services producers (like CloudSQL, Firebase, etc) in private subnet of the VPC. See https://cloud.google.com/vpc/docs/configure-private-services-access#allocating-range. See https://cloud.google.com/sql/docs/mysql/configure-private-services-access#configure-access."
```
Comment (mfarix) [inline variables.tf]: Maybe reword description to include the details here
https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#allocated_ip_range

Comment (cedric66) [reply variables.tf]: Added as requested

File: main.tf (hunk 8 above)
Comment (cedric66) [reply main.tf]: Added as requested

File: main.tf [hunk 10]
```diff
@@ -89,7 +91,6 @@ module "google_mysql_db" {
     retained_backups               = null
     retention_unit                 = null
   }
-
```
Comment (mfarix) [inline main.tf]: Keep the new line to show that below stuff are different

File: main.tf [hunk 11]
```diff
@@ -103,13 +104,16 @@ module "google_mysql_db" {
         ipv4_enabled        = var.public_access_read_replica
         private_network     = var.private_network
         require_ssl         = null
+        allocated_ip_range  = var.read_replica_pvt_ip_range
```
Comment (mfarix) [inline main.tf]: Can use same naming convention used above https://github.com/airasia/terraform-google-mysql_db/pull/10/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR81
Something like `allocated_ip_range_read_replica`

File: variables.tf [hunk 12]
```diff
@@ -140,12 +144,24 @@ variable "highly_available" {
...
+}
+
 variable "read_replica_count" {
   description = "Specify the number of read replicas for the MySQL instance. Value greater than 0 requires 'var.pit_recovery_enabled' to be 'true'."
   type        = number
   default     = 0
 }
 
+variable "read_replica_pvt_ip_range" {
+  description = "The name of the allocated ip range for the private ip CloudSQL instance. For example: google-managed-services-default. If set, the instance ip will be created in the allocated range. The range name must comply with RFC 1035. Specifically, the name must be 1-63 characters long and match the regular expression a-z?"
```
Comment (mfarix) [inline variables.tf]: May use similar description as `allocated_ip_range` variable. Also update the variable name as mention in above comment.

File: main.tf [hunk 13]
```diff
@@ -103,13 +104,16 @@ module "google_mysql_db" {
...
       }
-      database_flags      = local.db_flags_read_replica
-      disk_autoresize     = var.disk_auto_resize_read_replica
-      disk_size           = var.disk_size_gb_read_replica
-      disk_type           = "PD_SSD"
-      user_labels         = var.labels_read_replica
-      encryption_key_name = null
+      database_flags        = local.db_flags_read_replica
+      disk_autoresize       = var.disk_auto_resize_read_replica
+      disk_autoresize_limit = var.disk_autoresize_limit
```
Comment (syedrakib) [inline main.tf]:
```suggestion
      disk_autoresize_limit = var.disk_autoresize_limit_read_replica
```

File: variables.tf [hunk 14]
```diff
@@ -140,12 +144,24 @@ variable "highly_available" {
   default     = false
 }
 
+variable "highly_available_read_replica" {
+  description = "Whether the MySQL instance should be highly available REGIONAL or ZONAL. Highly Available (HA) instances will automatically failover to another zone within the region if there is an outage of the primary zone. HA instances are recommended for production use-cases and increase cost"
```
Comment (syedrakib) [inline variables.tf]:
```suggestion
  description = "Whether the MySQL instance should be highly available (REGIONAL) or not (ZONAL). Highly Available (HA) instances will automatically failover to another zone within the region if there is an outage of the primary zone. HA instances are recommended for production use-cases and, therefore, increase cost"
```

File: variables.tf [hunk 15]
```diff
@@ -115,7 +115,11 @@ variable "disk_auto_resize_read_replica" {
   type        = bool
   default     = false
 }
-
+variable "disk_autoresize_limit" {
+  description = "Allows users to set a specific limit on how large the storage on their instance can automatically grow. The default value is zero, which means there is no limit and disk size can grow up to the maximum available storage for the instance tier. Applying the automatic disk increase limit does not cause any disruptions to your database workload."
```
Comment (syedrakib) [inline variables.tf]:
```suggestion
  description = "Allows users to set a specific limit on how large the storage on their instance can automatically grow up to. The default value is zero, which means there is no limit and disk size can grow up to the maximum available storage for the instance tier. Applying the automatic disk increase limit does not cause any disruptions to your database workload."
```

File: variables.tf (hunk 12 above)
Comment (syedrakib) [reply variables.tf]: Also, the description needs to be revised to be clear that this is for the READ-REPLICA instances and not for the master instances.

And also need to clarify, from where can the "name" of the allocated IP range be obtained from.

File: variables.tf [hunk 16]
```diff
@@ -257,7 +284,11 @@ variable "additional_users" {
   }))
   default = []
 }
-
+variable "encryption_key_name_read_replica" {
+  description = "Encryption key is required for replicas in different regions. For replicas in same region as master, set encryption_key_name = null"
+  type        = string
+  default     = null
+}
```
Comment (syedrakib) [inline variables.tf:298]:
```suggestion
}

```

File: variables.tf [hunk 17]
```diff
@@ -194,6 +210,17 @@ variable "public_access_master_instance" {
   default     = false
 }
 
+variable "allocated_ip_range" {
+  description = <<-EOT
+  The name of the allocated ip range for the private ip CloudSQL instance. 
+  For example: "google-managed-services-default". If set, the instance ip will be created in the allocated range. 
+  The range name must comply with RFC 1035. Specifically, the name must be 1-63 characters long and match the regular expression a-z?. 
+  If set, a CIDR range of /20 is advised.
```
Comment (syedrakib) [inline variables.tf:225]: The description is saying to provide the name of an allocated IP range which will get used by CloudSQL. Where is that name defined in to begin with? From where can we obtain that name? We should clarify that in the description.

File: main.tf (hunk 10 above)
Comment (cedric66) [reply main.tf]: done

File: main.tf (hunk 11 above)
Comment (cedric66) [reply main.tf]: done

File: variables.tf (hunk 12 above)
Comment (cedric66) [reply variables.tf]: done

File: main.tf (hunk 13 above)
Comment (cedric66) [reply main.tf]: done

File: variables.tf (hunk 14 above)
Comment (cedric66) [reply variables.tf]: done

File: variables.tf (hunk 15 above)
Comment (cedric66) [reply variables.tf]: done

File: variables.tf (hunk 16 above)
Comment (cedric66) [reply variables.tf:298]: done

File: variables.tf (hunk 17 above)
Comment (cedric66) [reply variables.tf:225]: done

**Review by syedrakib — DISMISSED** [review]

File: variables.tf [hunk 18]
```diff
@@ -140,12 +144,30 @@ variable "highly_available" {
...
   default     = 0
 }
 
+variable "allocated_ip_range_read_replica" {
+  description = <<-EOT
+  The name of the allocated ip range for the private ip CloudSQL read replica instance. 
+  If already available this can be derived from VPC -> Private Service Connection -> Name of the allocated range
+  For example: "google-managed-services-default". If set, the instance ip will be created in the allocated range. 
+  The range name must comply with RFC 1035. Specifically, the name must be 1-63 characters long and match the regular expression a-z?. 
+  If set, a CIDR range of /20 is advised.
```
Comment (syedrakib) [inline variables.tf]:
```suggestion
  The name of the allocated IP range for the private IP of read replicas of the CloudSQL instance. 
  If already available this can be derived from VPC -> Private Service Connection -> Name of the allocated range. For example: "google-managed-services-default".
  If set, the IP will be created within the allocated IP range. 
  The range name must comply with RFC 1035. Specifically, the name must be 1-63 characters long and match the regular expression a-z?. 
  If set, a CIDR range of /20 is advised.
```

File: variables.tf [hunk 19]
```diff
@@ -194,6 +216,18 @@ variable "public_access_master_instance" {
   default     = false
 }
 
+variable "allocated_ip_range" {
+  description = <<-EOT
+  The name of the allocated ip range for the private ip CloudSQL instance. 
+  If already available this can be derived from VPC -> Private Service Connection -> Name of the allocated range
+  For example: "google-managed-services-default". If set, the instance ip will be created in the allocated range. 
+  The range name must comply with RFC 1035. Specifically, the name must be 1-63 characters long and match the regular expression a-z?. 
+  If set, a CIDR range of /20 is advised.
```
Comment (syedrakib) [inline variables.tf]:
```suggestion
  The name of the allocated IP range for the private IP of the CloudSQL instance. 
  If already available this can be derived from VPC -> Private Service Connection -> Name of the allocated range. For example: "google-managed-services-default".
  If set, the IP will be created within the allocated IP range. 
  The range name must comply with RFC 1035. Specifically, the name must be 1-63 characters long and match the regular expression a-z?. 
  If set, a CIDR range of /20 is advised.
```

**Review by syedrakib — COMMENTED** [review]
Some wording changes in the variable descriptions

**Review by syedrakib — APPROVED** [review]

**Review by mfarix — APPROVED** [review]
